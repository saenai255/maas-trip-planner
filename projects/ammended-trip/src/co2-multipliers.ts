enum TransportationMode {
    Airplane = 'AIRPLANE',
    Bicycle = 'BICYCLE',
    BicycleRent = 'BICYCLE_RENT',
    Bus = 'BUS',
    Car = 'CAR',
    EScooter = 'ESCOOTER',
    Ferry = 'FERRY',
    Funicular = 'FUNICULAR',
    Metro = 'METRO',
    PublicTransit = 'PUBLIC_TRANSIT',
    Rail = 'RAIL',
    Scooter = 'SCOOTER',
    SharedBicycle = 'SHARED_BICYCLE',
    Subway = 'SUBWAY',
    Taxi = 'TAXI',
    Tram = 'TRAM',
    Walk = 'WALK',
    Wait = 'WAIT',
    Train = 'TRAIN',
}

/**
 * CO2 Multipliers in g/km
 */
const CO2Multipliers: Record<TransportationMode, number> = {
    [TransportationMode.Airplane]: 144,
    [TransportationMode.Bicycle]: 0,
    [TransportationMode.BicycleRent]: 0,
    [TransportationMode.Bus]: 74,
    [TransportationMode.Car]: 160,
    [TransportationMode.EScooter]: 110,
    [TransportationMode.Ferry]: 144,
    [TransportationMode.Funicular]: 54,
    [TransportationMode.Metro]: 54,
    [TransportationMode.PublicTransit]: 74,
    [TransportationMode.Rail]: 14,
    [TransportationMode.Scooter]: 75,
    [TransportationMode.SharedBicycle]: 0,
    [TransportationMode.Subway]: 74,
    [TransportationMode.Taxi]: 250,
    [TransportationMode.Train]: 14,
    [TransportationMode.Tram]: 54,
    [TransportationMode.Walk]: 0,
    [TransportationMode.Wait]: 0,
} as const;

/**
 * Returns the CO2/km multiplier for the given transportation mode.
 */
function getCO2MultiplierOrNull(mode: TransportationMode): number | null {
    return CO2Multipliers[mode] ?? null;
}

/**
 * Returns the CO2 emissions (as grams) for a given distance and transportation mode.
 */
function calculateCO2Emissions(distanceInMeters: number, transportationMode: TransportationMode): number {
    const co2Multiplier = getCO2MultiplierOrNull(transportationMode);
    if (co2Multiplier === null) {
        throw new Error(`Could not find CO2 multiplier for transportation mode ${transportationMode}`);
    }

    const distanceInKm = distanceInMeters / 1000;
    return distanceInKm * co2Multiplier;
}

export { TransportationMode, getCO2MultiplierOrNull, calculateCO2Emissions };
