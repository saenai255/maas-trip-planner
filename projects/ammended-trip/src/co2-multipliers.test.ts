import { TransportationMode, calculateCO2Emissions, getCO2MultiplierOrNull } from './co2-multipliers';

describe('co2-multipliers', () => {
    describe('getCO2MultiplierOrNull', () => {
        it('should return a multiplier for a given mode', () => {
            expect(getCO2MultiplierOrNull(TransportationMode.Airplane)).toEqual(144);
        });

        it('should return null for an unknown mode', () => {
            expect(getCO2MultiplierOrNull('UNKNOWN' as TransportationMode)).toEqual(null);
        });
    });

    describe('calculateCO2Emissions', () => {
        it('should calculate the CO2 emissions', () => {
            expect(calculateCO2Emissions(1000, TransportationMode.Bus)).toEqual(74);
        });

        it('should throw when given an unknown mode', () => {
            expect(() => calculateCO2Emissions(1000, 'UNKNOWN' as TransportationMode)).toThrowErrorMatchingSnapshot();
        });
    });
});
