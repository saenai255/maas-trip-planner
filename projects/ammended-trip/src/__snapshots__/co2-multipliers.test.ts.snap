// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`co2-multipliers calculateCO2Emissions should throw when given an unknown mode 1`] = `"Could not find CO2 multiplier for transportation mode UNKNOWN"`;
